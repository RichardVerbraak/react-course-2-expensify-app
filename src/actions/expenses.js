import database from '../firebase/firebase'

// With firebase (is now asynchronous)
// component calls action generator
// action generator returns function
// component dispatches function (?)
// function runs internally by redux and then has the ability to dispatch other actions 

// ADD_EXPENSE
export const addExpense = (expense) => {
    return {
        type: 'ADD_EXPENSE',
        expense: expense
    }
}

// HARDER THEN IT LOOKS
// This dispatches another dispatch call AFTER it's been stored in firebase upon resolve
// We get an expense passed from the component to this action
// If there isn't one we set it default to an empty object
// Then we start off by creating a function that has access to dispatch (now possible too thunk for creating functions inside of actions)
// If there is one with empty fields we set the expenseData to default values
// We setup to destructure things from the expense we got from the component, and store in firebase AND redux store
// We push on the expense and save it to firebase and THEN, if resolved (promise/asynchronous) we dispatch the action to change the redux store
// .then has access to the reference (because it's now stored in firebase) and we store the firebase id in the redux store as well as the rest
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        // Default values if fields were left empty
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData

        const expense = { description, note, amount, createdAt }

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => {
    return {
        type: 'REMOVE_EXPENSE',
        id: id
    }
}

// Gets the id passed in from component --> removes it in the database --> then upon resolve, dispatches the removeExpense action to remove it from redux store
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({ id }))
        })
    }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => {
    return {
        type: 'EDIT_EXPENSE',
        id: id,
        updates: updates
    }
}

// Returns a function that dispatches editExpense only when firebase has stored the data
export const startEditExpense = (id, updates) => {
    return (dispatch) => {        
        return database.ref(`expenses/${id}`).update(updates)
        .then(() => {
            dispatch(editExpense(id, updates))
        })
    }
}

// SET_EXPENSES
export const setExpenses = (expenses) => {
    return {
        type: 'SET_EXPENSES',
        expenses
    }
}

// I previously couldn't see the new expenses loading on the page because createdAt was zero aka 1 Jan 1970
// I added another .then call after we pushed onto the expenses with forEach but its unecessary, it did still work
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value')
        .then((snapshot) => {  
            const expenses = []                           
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(expenses))
        })       
    }
}