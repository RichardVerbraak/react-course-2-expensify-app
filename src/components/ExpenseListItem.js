import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";
import { dutchLocale } from "../../locales/numeral-locale";

// Not 100% sure how to use numeral.js
numeral.locale("nl-nl");

// It doesn't have access to expenses object here but it does inside ExpenseList
// The props are already 'inside' of this component as well
// Whenever you use connect to connect a component to the Redux store, it'll inject dispatch as a prop to the component
const ExpenseListItem = ({ id, description, amount, createdAt }) => {
  return (
    <Link className="list-item" to={`edit/${id}`}>
      <div>
        <h3 className="list-item__title">{description}</h3>
        <span className="list-item__subtitle">{moment(createdAt).format("MMMM Do, YYYY")}</span>
      </div>
      <h3 className="list-item__amount">{numeral(amount / 100).format("$0,0.00")}</h3>
    </Link>
  );
};

// No need for mapStateToProps if you don't need state

export default ExpenseListItem;
