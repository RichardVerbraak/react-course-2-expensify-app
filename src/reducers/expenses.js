const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch(action.type) {
        // Gets state and returns new arr with the old + the new expense
        case 'ADD_EXPENSE':
            return [...state, action.expense]
        // Makes a new array except for the ones that didn't pass the test (so it excludes the one that did match)
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => {
                return expense.id !== action.id
            })
        // Map -> creates new arr with the results of the below expression that is done on every element
        // Overrides the old expense array with updated expense
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense
                }
            })
        default:
            return state
    }
}

export default expensesReducer