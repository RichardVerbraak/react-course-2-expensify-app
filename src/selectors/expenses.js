import moment from 'moment'

// Gets the filtered and sorted expenses
// Also sorts by the oldest created expense first or highest amount (?)
// Filters startDates out, based on if they are made on the same day or before the createdAt
// Expenses array gets filtered on the booleans coming from the ternary operator comparisons
// If there is no startDate we won't filter based of off that so we just use true (so it doesnt get filtered out or w/e)
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        // Is 1000 < 0 ? if not : put that one first (most recent)
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })    
}

export default getVisibleExpenses