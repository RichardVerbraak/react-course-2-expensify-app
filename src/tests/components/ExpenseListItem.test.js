import React from 'react'
import { Link } from 'react-router-dom'
import ExpenseListItem from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses'
import { shallow } from 'enzyme'

// We have to spread out expenses because we are destructuring them in ExpenseListItem
// It's the same as writing description={expenses[0].description} amount={expenses[0].amount} etc.
// A shorthand way of spreading all the properties on an object to props without having to manually type out every single one.
test('Should render first expense from fixture data', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>)
    expect(wrapper).toMatchSnapshot()
})