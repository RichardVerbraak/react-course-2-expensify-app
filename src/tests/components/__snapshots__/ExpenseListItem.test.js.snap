// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render first expense from fixture data 1`] = `
<Link
  className="list-item"
  to="edit/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Gum
    </h3>
    <span
      className="list-item__subtitle"
    >
      January 1st, 1970
    </span>
  </div>
  <h3
    className="list-item__amount"
  >
    â‚¬ 1,95
  </h3>
</Link>
`;
